crates/handlers/src/oauth2/end_session.rs

```diff
@@ -0,0 +1,211 @@
+        Self::UnknownToken
+    }
+}
+
+#[tracing::instrument(name = "handlers.oauth2.end_session.get", skip_all)]
+pub(crate) async fn get(
+    mut rng: BoxRng,
+    clock: BoxClock,
+    State(key_store): State<Keystore>,
+    State(url_builder): State<UrlBuilder>,
+    mut repo: BoxRepository,
+    activity_tracker: BoundActivityTracker,
+    Query(params): Query<EndSessionParam>,
+    cookie_jar: CookieJar,
+) -> Result<Response, RouteError> {
+    let (session_info, cookie_jar) = cookie_jar.session_info();
+
+    let browser_session_id = session_info
+        .current_session_id()
+        .ok_or(RouteError::BadRequest)?;
+
+    let browser_session = repo
+        .browser_session()
+        .lookup(browser_session_id)
+        .await?
+        .ok_or(RouteError::BadRequest)?;
+
+    let oauth_session = repo
+        .oauth2_session()
+        .find_by_browser_session(browser_session.id)
+        .await?
+        .ok_or(RouteError::BadRequest)?;
+
+    let client = repo
+        .oauth2_client()
+        .lookup(oauth_session.client_id)
+        .await?
+        .filter(|client| client.id_token_signed_response_alg.is_some())
+        .ok_or(RouteError::ClientNotFound)?;
+
+    let jwks = key_store.public_jwks();
+    let issuer: String = url_builder.oidc_issuer().into();
+
+    let id_token_verification_data = JwtVerificationData {
+        issuer: Some(&issuer),
+        jwks: &jwks,
+        signing_algorithm: &client.id_token_signed_response_alg.unwrap(),
+        client_id: &client.client_id,
+    };
+
+    verify_id_token(
+        &params.id_token_hint,
+        id_token_verification_data,
+        None,
+        clock.now(),
+    )?;
+
+    // Check that the session is still valid.
+    if !oauth_session.is_valid() {
+        // If the session is not valid, we redirect to post logout uri
+        return Ok((cookie_jar, Redirect::to(&params.post_logout_redirect_uri)).into_response());
+    }
+
+    // Check that the client ending the session is the same as the client that
+    // created it.
+    if client.id != oauth_session.client_id {
+        return Err(RouteError::UnauthorizedClient);
+    }
+
+    activity_tracker
+        .record_oauth2_session(&clock, &oauth_session)
+        .await;
+
+    // If the session is associated with a user, make sure we schedule a device
+    // deletion job for all the devices associated with the session.
+    if let Some(user_id) = oauth_session.user_id {
+        // Fetch the user
+        let user = repo
+            .user()
+            .lookup(user_id)
+            .await?
+            .ok_or(RouteError::UnknownToken)?;
+
+        // Schedule a job to sync the devices of the user with the homeserver
+        repo.queue_job()
+            .schedule_job(&mut rng, &clock, SyncDevicesJob::new(&user))
+            .await?;
+    }
+
+    // Now that we checked everything, we can end the session.
+    repo.oauth2_session().finish(&clock, oauth_session).await?;
```

Comment:
Not sure about that semantic. I don't know whether we should finish the oauth session at this point or let the client still call the revocation endpoint.

Either way, we need a better way to find the right OAuth session, and I think this should be done by injecting the OAuth session ID in the id_token `sid` claim when we give out the id_token